src/main/kotlin/org/ageseries/libage/sim/thermal/Thermal.kt

```diff
@@ -0,0 +1,306 @@
+    val contactPoint: Double = 0.5,
+) {
+    companion object {
+        val DEFAULT = ThermalConnectionParameters()
+    }
+}
+
+class ThermalConnection(
+    /** One of the connected masses. */
+    val a: ThermalMass,
+    /** The other connected mass. */
+    val b: ThermalMass,
+    /** Thermal parameters of this connection. */
+    params: ThermalConnectionParameters = ThermalConnectionParameters.DEFAULT,
+) {
+    val contactPoint = params.contactPoint
+    val distance = params.distance
+    val conductance = params.conductance
+
+    override fun toString() = "<Conn $a $b ${distance}m($contactPoint) ${conductance}W/K>"
+    /**
+     * Returns the change in energy, added to [a] and subtracted from [b], which would equilibriate the connected thermal masses over the given period of time [dt] (in s).
+     *
+     * For this to be stable, [dt] must be held relatively small, since this is a linear approximation to an exponential curve--otherwise, overshoot may be observed.
+     */
+    fun transfer(dt: Double): Double {
+        // Kelvin
+        val delta_T = b.temperature.kelvin - a.temperature.kelvin
+        // W/K
+        val dist_cond_a = a.material.conductance * contactPoint * distance
+        val dist_cond_b = b.material.conductance * (1.0 - contactPoint) * distance
+        val overall_cond = (dist_cond_a * dist_cond_b * conductance).pow(1.0/3.0)
+        // W
+        val power = delta_T * overall_cond
+        // J
+        return power * dt
+    }
+}
+
+interface ThermalBody<Locator> {
+    val mass: ThermalMass
+    val locator: Locator
+    /** Surface area of this body w.r.t. the [Environment], nominally in m^2. */
+    val surfaceArea: Double
+}
+
+interface Environment<Locator> {
+    /** The [Temperature] of the environment at this locator. */
+    fun temperature(locator: Locator): Temperature
+    /** How conductive the substance in the environment is at this locator. */
+    fun conductance(locator: Locator): Double
+}
+
+class Simulator<Locator>(val environment: Environment<Locator>) {
+    class Connection<Locator>(
+        val a: ThermalBody<Locator>,
+        val b: ThermalBody<Locator>,
+        params: ThermalConnectionParameters,
+    ) {
+        val connection = ThermalConnection(a.mass, b.mass, params)
+    }
+
+    val connections = mutableSetOf<Connection<Locator>>()
+    val bodies = mutableSetOf<ThermalBody<Locator>>()
+    val connectionMap = mutableMultiMapOf<ThermalBody<Locator>, Connection<Locator>>()
+
+    /**
+     * Add some bodies to the simulation.
+     *
+     * Without connection, these bodies will still conduct from/to the [environment].
+     */
+    fun add(vararg bodies_added: ThermalBody<Locator>) {
+        bodies.addAll(bodies_added)
+    }
+
+    /**
+     * Add some connections to the simulation.
+     *
+     * Generally, you would use [connect] instead.
+     */
+    fun add(vararg connections_added: Connection<Locator>) {
```

Comment:
`connectionsAdded`

Also, varargs in Kotlin are terribly nonperformant and you should use a list or something. I'd scan the whole repo for these and replace them.